addons/portal/controllers/portal.py

```diff
@@ -230,6 +232,325 @@ def account(self, redirect=None, **post):
+        """
+        PartnerSudo = request.env['res.partner'].with_context(show_address=1).sudo()
+        partner_sudo = PartnerSudo.browse(partner_id).exists()
+        if partner_sudo and not partner_sudo._can_be_edited_by_current_partner(address_type=address_type, **kwargs):
+            raise Forbidden()
+        return partner_sudo
+
+    def _prepare_address_form_values(self, partner_id, address_type, **_kwargs):
+        """ Prepare and return the values to use to render the address form.
+        It returns the sudoed partner editing the form and the values used in the form.
+
+        :param int partner_id: id of the partner editing the address form
+        :param str address_type: The type of the address: 'billing' or 'delivery'.
+        :param str template: rendering template to use
+        : param query_params: additional parameters
+        :rtype: tuple[record of `res.partner`, dict]
+        """
+        partner_id = partner_id and int(partner_id)
+        partner_sudo = self._check_partner_edit_rights(partner_id=partner_id, address_type=address_type, **_kwargs)
+        can_edit_vat = (
+            address_type in ['invoice', 'billing']
+            and (not partner_sudo or partner_sudo.can_edit_vat())
+        )
+        can_edit_vat = partner_sudo.type in ['invoice', 'billing'] and (not partner_sudo or partner_sudo.can_edit_vat())
+
+        ResCountrySudo = request.env['res.country'].sudo()
+        country_sudo = partner_sudo.country_id
+        if not country_sudo:
+            country_sudo = request.env.user.country_id
+
+        state_id = partner_sudo.state_id.id
+        address_fields = country_sudo and country_sudo.get_address_fields() or ['city', 'zip']
+        rendring_values = {
+            'partner_sudo': partner_sudo,  # If set, customer is editing an existing address
+            'partner_id': partner_sudo.id,
+            'address_type': address_type,  # 'billing' or 'delivery'
+            'type': address_type,
+            'can_edit_vat': can_edit_vat,
+            'discard_url': '/my/addresses',
+            'country': country_sudo,
+            'countries': ResCountrySudo.search([]),
+            'has_invoice': partner_sudo.can_edit_info() if partner_sudo else True,
+            'state_id': state_id or 1,
+            'country_states': country_sudo.state_ids,
+            'zip_before_city': (
+                'zip' in address_fields
+                and address_fields.index('zip') < address_fields.index('city')
+            ),
+            'show_vat': bool(address_type in ['contact', 'billing', 'invoice'] and not partner_sudo.parent_id),
+            'vat_label': _("VAT"),
+        }
+        return partner_sudo, rendring_values
+
+    @route(['/portal/country_info/<model("res.country"):country>'], type='json', auth="public", methods=['POST'], website=True)
+    def shop_country_info(self, country, address_type, **kw):
+        address_fields = country.get_address_fields()
```

Comment:
:warning: `get_address_fields()` uses `re.findall()` which can return empty matches. This way in line 331 `address_fields.index('zip') < address_fields.index('city')` could compare `None` and trigger a `TypeError`. I think we should use a default value as you did in the function above at:

`address_fields = country_sudo and country_sudo.get_address_fields() or ['city', 'zip']`